rep-2011.rst

```diff
@@ -0,0 +1,1421 @@
+
+When using DDS-Xtypes type compatibility is determined through sophisticated and configurable rules, allowing for things like extensible types, optional fields, implicit conversions, and even inheritance.
+Which of these features is supported for use with ROS is out of scope with this REP, but if any of them are in use, then it may be possible for two endpoints to match and communicate even if their ROS 2 type version hashes do not match.
+
+In this situation the middleware is responsible for communicating to the rmw layer when an endpoint will not be matched due to type incompatibility.
+The ``INCONSISTENT_TOPIC`` event in DDS applies for DDS-XTypes as well, and should be useful in fulfilling this requirement.
+
+Type Description Distribution
+-----------------------------
+
+For some use cases the type version hash is insufficient and instead the full type description is required.
+
+One of those use cases, which is also described in this REP, is "Run-Time Interface Reflection", which is the ability to introspect the contents of a message at runtime when the description for that message, or that version of that message, was unavailable at compile time.
+In this use case the type description is used to interpret the serialized data dynamically.
+
+Another use case, which is not covered in this REP, is using the type description in tooling to either display the type description to the user or to include it in recordings.
+
+In either case, where the type description comes from doesn't really matter, and so, for example, it could be looked up on the local filesystem or read from a rosbag file.
+However, in practice, the correct type description may not be found locally, especially in cases where you have different versions of messages in the same system, e.g.:
+
+- because it's on another computer, or
+- because it is from a different distribution of ROS, or
+- because it was built in a different workspace, or
+- because the application has not been restarted since recompiling a change to the type being used
+
+In any case, it is useful to have a mechanism to convey the type descriptions from the source of the data to other nodes, which we describe here as "type description distribution".
+
+Furthermore, this feature should be agnostic to the underlying middleware and serialization library, as two endpoints may not have the same rmw implementation, or the data may have been serialized to a different format in the case of playback of a recording.
+
+Sending the Type Description
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Type descriptions will be provided by a ROS Service called ``~/_get_type_description``, which will be offered by each node.
+There will be a single ROS Service per node, regardless of the number of publishers or subscriptions on that node.
+
+This service may be optional, for example being enabled or disabled when creating the ROS Node.
+
+.. TODO::
+
+    How can we detect when a remote node is not offering this service?
```

Comment:
Just agreeing that yes this is challenging. `rosbag2` is going to be a first adopter of these services, and have a strong preference for finding type descriptions. How will it know when to stop checking? Like you've pointed out, hard to tell whether not ready yet, or never will be.